core/src/main/java/org/apache/calcite/rel/metadata/RelMetadataQuery.java

```diff
@@ -75,8 +76,17 @@
  * plugin mechanism.
  */
 public class RelMetadataQuery {
-  /** Set of active metadata queries, and cache of previous results. */
-  public final Map<List, Object> map = new HashMap<>();
+  /**
+   * Set of active metadata queries, and cache of previous results.
+   *
+   * Can be sealed when no more additions to the map are expected.
+   * Write operations on the map will fail after sealing
+   *
+   * Sealing is helpful if no additions to the map are expected or desired.
+   * The map here is unbounded so it can leak memory if not properly cleaned
+   * up
+   */
+  public final SealableMap<List, Object> map = new SealableMap<>(new HashMap<>());
```

Comment:
Who will call `map.seal()`? and when?

Reply:
Needs to be called whenever no more additions to this map are expected (which is usually when no more RelNodes are expected to be added to the RelOptCluster)

Reply:
As for the who -- it really depends on the structure of the code but I would expect whatever code holds RelNodes for a long time would try to do this (`relNode.getCluster().getRelMetadataQuery().map.seal()`)

Reply:
This is not good encapsulation from Calcite code. You can ask a sealCache() method to RelMetadataQuery instead.
So we will do `relNode.getCluster().getRelMetadataQuery().sealCache()` instead.